book-2-mvc/chapters/USER_RELATED_DATA.md

```diff
@@ -0,0 +1,295 @@
+
+            List<Dog> dogs = new List<Dog>();
+
+            while (reader.Read())
+            {
+                Dog dog = new Dog()
+                {
+                    Id = reader.GetInt32(reader.GetOrdinal("Id")),
+                    Name = reader.GetString(reader.GetOrdinal("Name")),
+                    Breed = reader.GetString(reader.GetOrdinal("Breed")),
+                    OwnerId = reader.GetInt32(reader.GetOrdinal("OwnerId"))
+                };
+
+                // Check if optional columns are null
+                if (reader.IsDBNull(reader.GetOrdinal("Notes")) == false)
+                {
+                    dog.Notes = reader.GetString(reader.GetOrdinal("Notes"));
+                }
+                if (reader.IsDBNull(reader.GetOrdinal("ImageUrl")) == false)
+                {
+                    dog.ImageUrl = reader.GetString(reader.GetOrdinal("Notes"));
+                }
+
+                dogs.Add(dog);
+            }
+
+            reader.Close();
+
+            return View(dogs);
+        }
+    }
+}
+```
+
+Now try going to the `/dogs` route. You should only see the dogs of the owner you just logged in as.
+
+### Defaulting the OwnerId when creating dogs
+
+Currently the form the user fills out when creating a dog asks the user to fill out a field for OwnerId. We can now remove that field and have the server set that value for us. First, go into the `Create.cshtml` file for Dogs and remove the form field for `OwnerId`
+
+Now change the controller method for `Create` to set the `OwnerId` to the current user's Id
+
+```csharp
+[HttpPost]
+[ValidateAntiForgeryToken]
+public ActionResult Create(Dog dog)
+{
+    try
+    {
+        using (SqlConnection conn = Connection)
+        {
+            conn.Open();
+            using(SqlCommand cmd = conn.CreateCommand())
+            {
+                cmd.CommandText = @"
+                    INSERT INTO Dog (Name, Breed, OwnerId, Notes, ImageUrl)
+                    VALUES (@name, @breed, @ownerId, @notes, @imageUrl)
+                ";
+
+                cmd.Parameters.AddWithValue("@name", dog.Name);
+                cmd.Parameters.AddWithValue("@breed", dog.Breed);
+                cmd.Parameters.AddWithValue("@ownerId", GetCurrentUserId());
+
+                // nullable columns
+                cmd.Parameters.AddWithValue("@notes", dog.Notes ?? "");
+                cmd.Parameters.AddWithValue("@imageUrl", dog.ImageUrl ?? "");
+
+                cmd.ExecuteNonQuery();
+
+                return RedirectToAction(nameof(Index));
+
+            }
+        }
+
+    }
+    catch (Exception ex)
+    {
+        return View(dog);
+    }
+}
+```
+
+## Exercise
+
+Update the Index method in the walkers controller so that owners only see walkers in their own neighborhood. 
+
+**Hint**: you can get the logged in owner's neighborhood Id like this
+
+```csharp
+string neighborhoodIdString = User.FindFirstValue("NeighborhoodId");
```

Comment:
This may sound odd, but I think this forces them to understand cookies and claims more than we need them to.